### DISCUSSION TITLE
[cgen, checker, pref]: Streamline architectures

### DISCUSSION THREAD

**Discussion by IngwiePhoenix** [vlang/v#11840] [Ideas]
While porting V to the GBA via the DevkitPRO toolchain, I noticed that architectures are a little bit wobbly right now. Since I am not one to make decisions and don't personally have a lot of knowledge about the diverse archs out there, I'd rather request this as a feature instead and let more knowledged people handle this.

* Should the various ARM versions (ARMv4 - ARMv8 and friends) be defined separately or just as one?
* Should the `__ARM_ARCH` define be respected and used?
* Should idents be `aarch16`, `aarch32` and `aarch64` OR called by their code names respectively? Like `arm_thumb`
* checker and cgen: Checker might allow comptime idents that the cgen might not know about - and vice versa. These should possibly be synced.

Since `v/pref/os.v` already exists with methods to convert to and from strings, should an `arch.v` be added to take care of handling of architectures? This might also help `v/table/types.v` to pick the correct `Language` - which I assume is used for `asm $arch { ... }` statements.

Aside from that, I have not found any mention of MIPS (PSP, PS Vita and many other devices - but those are the ones I am most familiar with). RISC-V seems to be well covered already.

And finally - should `wasm32` and `wasm64` be added as well? Although there is no actual WASM backend just yet, it might become useful at a later point and could already be considered whilst streamlinign architectures.

Let me know what you think! :)

**Comment by JalonSolov** [comment]
The `aarch` name is a shortening of `arm architecture`.  Whether `aarch` or `arm` is used may depend on whether we have implicit vs explicit targets for each individual variation of an ARM CPU that's out there.

As for `wasm32` and `wasm64`... I'd vote to wait until we have something before we add these.  It would be just what V needs if someone saw they were there, then wrote an article on HN complaining about V lying again claiming something that didn't work.  :-\
